include/dlaf/eigensolver/tridiag_solver/merge.h

```diff
@@ -961,135 +970,449 @@ void reduceSumScalingVector(common::Pipeline<comm::Communicator>& col_task_chain
+              const SizeType j_subm_el = i2_perm[i_subm_el];
+
+              // if it is a deflated vector
+              if (j_subm_el >= k) {
+                const GlobalElementIndex ij_el(origin_el.rows() + i_subm_el,
+                                               origin_el.cols() + j_subm_el);
+                const GlobalTileIndex ij = dist.globalTileIndex(ij_el);
+
+                if (dist.rankIndex() == dist.rankGlobalTile(ij)) {
+                  const LocalTileIndex ij_lc = dist.localTileIndex(ij);
+                  const SizeType linear_subm_lc =
+                      (ij_lc.row() - ij_begin_lc.row()) +
+                      (ij_lc.col() - ij_begin_lc.col()) * sz_loc_tiles.rows();
+                  const TileElementIndex ij_el_tl = dist.tileElementIndex(ij_el);
+                  evec_tiles[to_sizet(linear_subm_lc)](ij_el_tl) = T{1};
+                }
+              }
+            }
           }
         }
-      }
-    }
 
-    // Fill ones for deflated Eigenvectors.
-    // Quick return if there is none.
-    if (n == k)
-      return;
-
-    const GlobalElementIndex origin{i_begin * dist.blockSize().rows(),
-                                    i_begin * dist.blockSize().cols()};
-    for (SizeType i = 0; i < n; ++i) {
-      const SizeType j = i2_ptr[i];
-      if (j >= k) {
-        const GlobalElementIndex i_g{i + origin.row(), j + origin.col()};
-        const GlobalTileIndex i_tile = dist.globalTileIndex(i_g);
-        if (dist.rankIndex() == dist.rankGlobalTile(i_tile)) {
-          const LocalTileIndex i_tile_l = dist.localTileIndex(i_tile);
-          const SizeType i_subm_evec_arr = i_tile_l.row() - idx_loc_begin.row() +
-                                           (i_tile_l.col() - idx_loc_begin.col()) * sz_loc_tiles.rows();
-          const TileElementIndex i = dist.tileElementIndex(i_g);
-          evec_tile_arr[to_sizet(i_subm_evec_arr)](i) = T{1};
+        // STEP 0b: Initialize workspaces (single-thread)
+        if (thread_idx == 0) {
+          // Note:
+          // - nthreads are used for both LAED4 and weight calculation (one per worker thread)
+          // - last one is used for reducing weights from all workers
+          ws_cols.reserve(nthreads + 1);
+
+          // Note:
+          // Considering that
+          // - LAED4 requires working on k elements
+          // - Weight computaiton requires working on m_subm_el_lc
+          //
+          // and they are needed at two steps that cannot happen in parallel, we opted for allocating the
+          // workspace with the highest requirement of memory, and reuse them for both steps.
+          const SizeType max_size = std::max(k, m_subm_el_lc);
+          for (std::size_t i = 0; i < nthreads; ++i)
+            ws_cols.emplace_back(max_size);
+          ws_cols.emplace_back(m_subm_el_lc);
+
+          ws_row = memory::MemoryView<T, Device::CPU>(n_subm_el_lc);
+          std::fill_n(ws_row(), n_subm_el_lc, 0);
         }
```

Comment:
Yes, correct.

I didn't bother too much with trying to improve step0a because, as stated in the comment

```
// Note: this step is completely independent from the rest, but it is small and it is going
// to be dropped soon.
```